### PR TITLE
Fix: keep parens around `TSInstantiationExpression` followed by a prop access

### LINKED ISSUE #14625: Prettier removes parenthesis around Typescript instantiation expression that is used as part of a bind call
**Prettier 2.8.7**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAngeQCMArOMLAXk2AB0pNMBBJTARgBo6GAhFgJk71MAYRYBmOgF9MAQzSZIUDAG46dGLgAOcTAFkCJMjABqMgDYBXOABUtOqgAoN2iADMcB0uQCUAbQDWcLhumM5wIXhEXjAAuqpQdK4WUOQAltCYqWimZqkAJjlWADzGmHAAHvBQefJQFtiEcABOAHwOXJgQJCwASmQQTXlFGE2pUADm7JjGLVMdAG7mVix1Dc103iyLljpZ09QdTXAwFk30UUYAdNtWaA5dxN6XY2CWeXB3Nzpy097xkmooEkUjB0vRxsccvlCnASmVKggaphVo1Wu0hA9ev1BsMYKMJlMZoIGF8WCMxuNMAAfZH1VHU2lmMwM5LvVxjOB5DYsUo01lwdlQTkHISpdwOL6YCjSxnMqk0yXSqj8wWc7wihgMI4nM6YFUcrlCaRwMxoHS0IQMRQYWlrJowqWYABydOaEqWcD+HS1x1O9CyUIKHvuJCmKOaMPVAH5baiHSx9UK8vEGACoGm6NasBCTOZ8voLuQHY4c4GYUUC4Yix7bNoWk9CGM8g46kyppEqzA-iB2CAIJpQehkKAZE0mhAAO4ABVHCDQyBA5gnMlw897hCaMjAgRgAGUZNg4AAZDnIVzmM3rzfb467zRbinIPFWXtwNZ5d55I8yCYWGQQgAxAZsBkGBQQmBcZAsGAIB7EAAAsYGwMwAHV4NSeA0HvMA4F3OcMNSeYMNwBcwDQNcQDGM0mhgKdN3GECzwvOBe2INByl3CkzDgABFCwIHgJjTRYkB7yaaiFzCNAwFGAc4M0fEYBQ-IYHg5AAA4AAZewUiAzRQzdNAXBSPmaeY4DggBHfj4Do-t5xQOQAFohU5Tk4KOazUiOOj-0YpBz2E3szWwVInyaF8QDQLi4EYMDRkIaC4CnZoTyFITLyimK+IEiyAuY3sYBkQhlLyVTkAEEA8RkVJcgmYQIGwfyQA+ABWOCLDNaxiocwLMvmKwAElqgQPcZNSAdGGqXcNG4jK4EkSQgA)

```sh
--parser typescript
```

**Input:**

```tsx
const myObject = {
  A: 1,
  B: 2,
  C: 3
} as const;

type MyObjectValueType = (typeof myObject)[keyof typeof myObject];

function isValidValue<V extends number>(
  obj: Record<string, V>, 
  value: number
): value is V {
  return Object.values(obj).includes(value as V);
}

function getValidValue<V extends number>(
  obj: Record<string, V>,
  value: string | number | null | undefined
): V | undefined {
  if (value === null || value === undefined) {
    return undefined
  } else {
    const numberValue = Number(value);
    return isValidValue(obj, numberValue) ? numberValue : undefined;
  }
}

const getValidMyObjectValue = (getValidValue<MyObjectValueType>).bind(null, myObject);
```

**Output:**

```tsx
const myObject = {
  A: 1,
  B: 2,
  C: 3,
} as const;

type MyObjectValueType = (typeof myObject)[keyof typeof myObject];

function isValidValue<V extends number>(
  obj: Record<string, V>,
  value: number
): value is V {
  return Object.values(obj).includes(value as V);
}

function getValidValue<V extends number>(
  obj: Record<string, V>,
  value: string | number | null | undefined
): V | undefined {
  if (value === null || value === undefined) {
    return undefined;
  } else {
    const numberValue = Number(value);
    return isValidValue(obj, numberValue) ? numberValue : undefined;
  }
}

const getValidMyObjectValue = getValidValue<MyObjectValueType>.bind(
  null,
  myObject
);

```

**Expected behavior:**
The parenthesis should not be removed from the Typescript instantiation expression.
```tsx
const getValidMyObjectValue = (getValidValue<MyObjectValueType>).bind(
  null,
  myObject
);
```
If they are, it results in the following Typescript error
```sh
TS1477: An instantiation expression cannot be followed by a property access.
```

[Typescript playground for unformatted code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAtgTwPICMBWBTYsC8MDeAUDDAIIBcMAjADTEwBClATHSQMKUDMhAvjAEMIMUJCgBuQoSgIADhhgBZZOixQAagIA2AVwwAVOQrwAKGfJAAzeCszYAlAG0A1hgRWY5jB8So7UAF1JQksdMGwAS3AYCIhNLQiAE3i9AB51GAwADygMMEThMB04FAwAJwA+E3oQdEoAJSwQMsTU6DKIsABzGhh1Ct76ADdtPUoikvLCe0oR3QVYvoJ6MowoHTKwGD81ADo5vQgTWrR7Xc7gXUSMI4OFIT77SV4pUPCoKK2utfiklIx0pkcnkCjAJqVKtUSCcGk0Wm0oB1ur1+mwYHdKO1Ol0YAAfMHFCF4glaLTEsLXSydDCJaaUDL4ikYKlgGnLEgRawmO4wHB8klk3H4nl8vBMlk0+zskgwVbrTYwcXU2kkfgYLQQBREGUicDQAmTMr-XkwAByhPK3NGGCe9BIco2W1iv2S1uO6F64PK-ylAH4DRDjZQlazEpJVXwpKJ9d8NNoksodthjaZYy7-qlE6pk9bDPIKmcUJ1EiYiqTer5s1AnkA)

[Typescript playground for formatted code](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAtgTwPICMBWBTYsC8MDeAUDDAIIBcMAjADTEwBClATHSQMKUDMdAvjAEMIMUJCgBuQoSgIADhhgBZZOixQAagIA2AVwwAVOQrwAKGfJAAzeCszYAlAG0A1hgRWY5jB8So7UAF1JQksdMGwAS3AYCIhNLQiAE3i9AB51GAwADygMMEThMB04FAwAJwA+E3oQdEoAJSwQMsTU6DKIsABzGhh1CrYYADdtPUoikvLCe0oR3QVYvoJ6MowoHTKwGD81ADo5vQgTWrR7Xc7gXUSMI4OFIT77SV4pUPCoKK2utfiklIx0pkcnkCjAJqVKtUSCcGk0Wm0oB1ur1+oM7pR2p0ujAAD5g4oQ3H4rRaIlha6WToYRLTSgZPHkjCUsDU5YkCLWEx3GA4XnE0k4vHc3l4RnM6n2NkkGCrdabGBiqmJSQkfgYLQQBREaUicDQfGTMr-HkwAByBPKXNGGCe9BIso2W1iv2S1uO6F64PK-0lAH4DRDjZRFSzlfQXi9CKJ9d8NNoksodthjXhYy7-qlE6pk9bDPIKrsUJ1ElD+YNfNmoNNxEA)

### PR DESCRIPTION
**Pull request by morsko1**
## Description

Fixes #14625 

Keep parentheses around `TSInstantiationExpression` followed by a property access.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -890,6 +890,12 @@ function needsParens(path, options) {
           parent.type !== "VariableDeclarator" &&
           parent.type !== "YieldExpression")
       );
+
+    case "TSInstantiationExpression":
+      return (
+        isMemberExpression(parent) &&
+        node.typeParameters?.type === "TSTypeParameterInstantiation"
```
Comment (fisker) [inline src/language-js/needs-parens.js:897]: MemberExpression has two children.

**Prettier pr-14701**
[Playground link](https://deploy-preview-14701--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAKAggJ0wQwJ4A8AzjJgJZQDmAfAJQB0OA3ADpQ4DaG2+xpFNWgF02IADQgIABxhloRZKBzYIAdwAKyhApQ4ANqvwKJAI1xgA1nBgBlHAFs4AGQpxkAM31E4p81dtSOGACyKQArj4g3vZkoZgREkQCenAAimEQ8B5ekQBWRAAeNslpGVlInnreEgCOZXDqmNI6IDhEALRQcHAAJr3iIKQ4ZHoCAMIQ9vY4yK16egNJVCnoMPwmYfDqcJguXdlVkQAWMPZ6AOpHZPBEgWBwNtrXZABu13izYETGIC8RAJJQPqwGxgcgydBAmwwPApA7VEBSJrec64KSzJFwbyYF5uCQUbEwRo4SjTeGRQKYbGzGFSLFgsgyAZIigwc5kHowI7IAAcAAYJJg4HUyELiaSZhUchIYDgTOzOdykAAmCRhbwAFTlOkqCLg9hMvT6PScOCoYRJcAAYhBMNM1iFdJsICAAL6uoA)

```sh
--parser typescript
```

**Input:**

```tsx
(Array<string>).a;
a[(Array<string>)]

```

**Output:**

```tsx
(Array<string>).a;
a[(Array<string>)];

```

Comment (morsko1) [reply src/language-js/needs-parens.js]: Thanks for suggestion.
I have updated the logic, so it keeps parens only if parent's property is accessed.
So, we keep parens if there is any parent's `property` object except `property` with type `TSInstantiationExpression`.

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -890,6 +890,13 @@ function needsParens(path, options) {
           parent.type !== "VariableDeclarator" &&
           parent.type !== "YieldExpression")
       );
+
+    case "TSInstantiationExpression":
+      return (
+        node.typeParameters?.type === "TSTypeParameterInstantiation" &&
+        parent.property?.type &&
+        parent.property.type !== "TSInstantiationExpression"
+      );
```
Comment (fisker) [inline src/language-js/needs-parens.js:899]:
```js
(Array<string>)[(Array<string>)]
```

Comment (fisker) [reply src/language-js/needs-parens.js:899]: The logic should be 


```js
key === "object" && isMemeberExpression(parent);
```

File: changelog_unreleased/typescript/14701.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+a[(Array<string>).name];
+(Array<string>).a;
+(Array<string>)?.a;
+(Array<string>)[a];
+(Array<string>)?.[a];
+(Array<string>)["a"];
+(Array<string>)?.["a"];
+(Array<string>)[`a`];
+(Array<string>)?.[`a`];
+(Array<string>)[(Array<string>)];
```
Comment (fisker) [inline changelog_unreleased/typescript/14701.md:19]: We don't need put everything in changelog, I think the following cases are enough.

```js
(Array<string>).a;
(Array<string>)?.a;
(Array<string>)[a];
(Array<string>)?.[a];
```

Comment (morsko1) [reply changelog_unreleased/typescript/14701.md]: updated

**Review by fisker — APPROVED** [review]
Thank you!

**Pull request merged**